### ISSUE TITLE
Question about optimal transport training

### ISSUE THREAD

**Issue by mli0603** [magicleap/SuperGluePretrainedNetwork#24]
Hi,

Thank you so much for sharing your pretrained network! This is really helpful 👍 

I am trying to reproduce the result with my own dataset. However, I face some problem where optimal transport seems to prohibit leanring of earlier layers. The loss will increase first and after a few iterations stuck at one place. If I train the early layers without optimal transport (keeping everything the same), I do not see such problem anymore. In the figure below, the **blue** curve is the training loss **with optimal transport layer** and the **red** curve is the training loss **without optimal transport**.

![image](https://user-images.githubusercontent.com/13513172/89688905-2c3a3c80-d8d1-11ea-8cfc-991d7ee5612e.png)

Have you encountered similar issues? Do you have any recommendation for training with optimal transport (such as learning rate etc)?

Thank you again!

**Comment by Skydes** [comment]
Hi!

1. What normalization do you apply to the similarity matrix when you don't use optimal transport?
2. Assuming that you used our released model as reference implementation (with Sinkhorn in log space, scaling of similarity matrix, weight initialization), a few points can go wrong:
    1. Learning rate is too high: 1e4 usually works well.
    2. Too few Sinkhorn iterations (<30) might affect the quality of the normalization.
    3. Imbalance between positive and negative examples: at the beginning of the training, only matchable points provide meaningful supervision, while unmatchable points (and the dustbins) behave like noise. As such, if your dataset has many more unmatchable points for each image pair, I recommend to balance the positive and the negative parts of the loss (1st vs 2nd+3rd terms of Equation 10). You can try to weight them equally, or even to put ~90% of the weight on the positive term (and maybe anneal throughout the training).

Hope this helps!

**Comment by mli0603** [comment]
@Skydes Thanks for your reply!

Regarding your comments
1. I use softmax if there is no optimal transport.
2. 

- I assume you meant 1e-4 instead of 1e4? Does the alpha (i.e. the appended value for dustbins) need such a large learning rate?
- I used 40 iterations. From what I can see, the optimal transport is doing the right thing (e.g. ensuring sum equal to marignal distribution and etc.), but the early layers are having a hard time to get well-initialized similarity matrix.
- I have attempted to remove the 2nd and 3rd terms entirely and only train positive examples just to test if matched points can be learnt. No luck there. I guess removing 2nd+3rd terms may not help because network doesn't know how to handle unmatched points? But this is a very valid point!

As a side note, as a post-processing tool (e.g. early layers have already finished training), I have optimal transport working as intended. That's why I am a little confused why training with optimal transport fails lol

**Comment by Skydes** [comment]
Oops, I meant 1e-4 indeed :) Om my datasets training with softmax and without the 2nd+3rd terms works just as well as training with the optimal transport layer when initially weighting positives much more. If you don't observe this, there might be a problem in your setup. You should monitor the recall, especially at the beginning of the training.

One last thing: initializing the dustbin score to a lower value, like 0, might also help. https://github.com/magicleap/SuperGluePretrainedNetwork/blob/master/models/superglue.py#L218

**Comment by mli0603** [comment]
@Skydes I see! I will try to weighing the positivies more and see if I can get a more reasonable result.

Thanks for your reply!
